### PR TITLE
Fixed Tar command when unpacking archive with missing intermediate directories

### PR DESCRIPTION
**Pull request by anti-social**

### CODE REVIEW

File: src/file_util.rs [hunk 1]
```diff
@@ -33,29 +33,82 @@ pub fn read_visible_entries(dir: &Path) -> Result<Vec<PathBuf>, Error> {
...
+        self
+    }
+
+    pub fn create(&self) -> Result<(), Error> {
+        create_dir(self.path.as_ref(), self.recursive,
+            self.mode, self.uid, self.gid)
+    }
+}
+
+pub fn create_dir(path: &Path, recursive: bool,
```
Comment (tailhook) [inline src/file_util.rs]: If you don't keep this function for backward compatibility (i.e. you change arguments anyway). then don't make it public anymore. If it's not public just make it a method and don't "unpack" structure items manually.

File: src/file_util.rs [hunk 2]
```diff
@@ -33,29 +33,82 @@ pub fn read_visible_entries(dir: &Path) -> Result<Vec<PathBuf>, Error> {
...
+    if uid.is_some() || gid.is_some() {
+        let uid = if let Some(uid) = uid {
+            uid
+        } else {
+            try!(path.symlink_metadata()).uid()
+        };
+        let gid = if let Some(gid) = gid {
+            gid
+        } else {
+            try!(path.symlink_metadata()).gid()
```
Comment (tailhook) [inline src/file_util.rs]: You make a "stat()" call twice in a row here. In some cases, like unpacking a large archive it can make a large difference.

Comment (anti-social) [inline src/file_util.rs]: Only once cause we enter in this block only if `uid.is_some() || gid.is_some()`.

Comment (tailhook) [inline src/file_util.rs]: Ah, yes I overlooked that.

**Comment by anti-social** [comment]
`.recursive_with_permissions(true)` method will be ok?

**Comment by anti-social** [comment]
Think `.propagate_permissions(true)` is better because `.recursive_with_permissions(false)` is ambigious.

**Comment by anti-social** [comment]
> On the other hand, if we have a !Snapshot which has owner enforced in configuration file we must create intermediate directories with that owner.

As I can see we don't create mount point directories. So at this moment propagating permissions is useless.

File: src/file_util.rs (hunk 1 above)
Comment (anti-social) [inline src/file_util.rs]: If I make it a method then it will be compiled for every generic `P: AsRef<Path>`.

**Comment by anti-social** [comment]
Updated.

**Comment by tailhook** [comment]
Merged. Thanks!

**Pull request merged**